### PR TITLE
auth: wire service account proxy

### PR DESCRIPTION
**Pull request by linoman**
**What is this feature?**

This PR will deregister the service account endpoints from the service manager and instead, register them through the HTTP server.

This PR will undo the work for [here](https://github.com/grafana/grafana/pull/58961).

**Why do we need this feature?**

Now that the external service account service is being added, we need to prevent modifications to the external service accounts created. By adding a proxy on top of the service, we can run validations on modifying these external service accounts.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by IevaVasiljeva — COMMENTED** [review]
The code changes look fine to me, but I think I don't fully understand the reasoning behind them. Do we already have a proxy to validate the requests? Did we consider validating service account modification in the handler for service account updates?

**Comment by kalleep** [comment]
I am not a big fan using wire to initiate api structures, I know it is done in other places but it has no value as a dependency to other services, i.e. it is not meant to be called. Why don't just the proxy wrapper just create and register the api, that seems like an easier fix. If I understand this correctly the proxy wrapper is meant to be used throughout the code base except in some instances where we want to allow creation of external service account and then I guess we would inject `*manager.ServiceAccountService` instead?

One problem for this approach is in those cases we cannot mock / fake this structure for tests

**Comment by linoman** [comment]
@IevaVasiljeva 

> Do we already have a proxy to validate the requests?
We developed the proxy but didn't wire it right away.

> Did we consider validating service account modification in the handler for service account updates?
Yes! The problem we were trying to solve was to protect external service accounts from everywhere, like API requests. We still want to be able to modify external service accounts from the external service account service.

**Comment by linoman** [comment]
> Why don't just the proxy wrapper just create and register the api, that seems like an easier fix.

I believe you mean to move the proxy/wrapper next to the registration at service level. The big problem with this is that we're trying to achieve 2 things.

1. Ensure any other package that uses the SA service interface is using the proxy, not only for API calls.
2. The External Service Account service, needs to be able to use the Service Account Service without the proxy.

**Comment by kalleep** [comment]
> > Why don't just the proxy wrapper just create and register the api, that seems like an easier fix.
> 
> I believe you mean to move the proxy/wrapper next to the registration at service level. The big problem with this is that we're trying to achieve 2 things.
> 
> 1. Ensure any other package that uses the SA service interface is using the proxy, not only for API calls.
> 2. The External Service Account service, needs to be able to use the Service Account Service without the proxy.

No not really. We still do some of the changes done in wire, binding `serviceaccount.Service` interface to the proxy.

**Comment by mgyongyosi** [comment]
> I am not a big fan using wire to initiate api structures, I know it is done in other places but it has no value as a dependency to other services, i.e. it is not meant to be called. Why don't just the proxy wrapper just create and register the api, that seems like an easier fix. If I understand this correctly the proxy wrapper is meant to be used throughout the code base except in some instances where we want to allow creation of external service account and then I guess we would inject `*manager.ServiceAccountService` instead?

I don't know what the current guidelines are around this (if there's any), but when we've done this with @linoman I suggested that the API should depend on the `serviceaccounts.Service` interface (depend on abstraction and not on concretion), rather than being instantiated by the manager. And we saw that other teams have already done the same (moving the Api as a dependency of `http_server.go`). However, I agree, probably it's better to register the Api in the service, because we won't implement other decorators on top of this.

**Comment by linoman** [comment]
I've updated the code following @kalleep suggestions

File: pkg/services/serviceaccounts/manager/service.go [hunk 1]
```diff
@@ -70,9 +69,6 @@ func ProvideServiceAccountsService(
 
 	usageStats.RegisterMetricsFunc(s.getUsageMetrics)
 
-	serviceaccountsAPI := api.NewServiceAccountsAPI(cfg, s, ac, accesscontrolService, routeRegister, permissionService)
```
Comment (gamab) [inline pkg/services/serviceaccounts/manager/service.go:73]: You can remove `permissionService`, `ac` and `routeRegister` from the arguments of `ProvideServiceAccountsService`.

**Review by gamab — APPROVED** [review]
LGTM! Good job. I only left one comment to clean up the `ProvideServiceAccountsService` arguments list now that it's not registering the API routes :slightly_smiling_face:

File: pkg/services/serviceaccounts/manager/service.go (hunk 1 above)
Comment (linoman) [reply pkg/services/serviceaccounts/manager/service.go:73]: Good catch! Dependencies removed

**Review by IevaVasiljeva — APPROVED** [review]
After clarifications from Lino and Gabriel it's a 🟢 from me!

**Review by mgyongyosi — APPROVED** [review]

**Review by kalleep — APPROVED** [review]

**Pull request merged**